API-doc-FORD-file.md

```diff
@@ -0,0 +1,51 @@
+---
+project: Fortran-lang/stdlib
+summary: A community driven standard library for (modern) Fortran
+src_dir: src
+exclude_dir: src/tests
+output_dir: API-doc
+page_dir: doc
+fpp_extensions: fypp
+preprocess: true
+macro: MAXRANK=4
+preprocessor: fypp
+display: public
+         protected
+         private
+source: true
+proc_internals: true
+md_extensions: markdown.extensions.toc
+graph: true
```

Comment:
It currently takes ~20 minutes to generate the graphs. We may decide to disable this. We could set this to false and then manually use the [entity metadata to enable a subset of graphs](https://github.com/Fortran-FOSS-Programmers/ford/wiki/Documentation-Meta-Data#graph).

Reply:
MAXRANK could be limited to 3. However, this will only marginally reduced the time. Because more procedures are expected, disabling the automatic generation of the graphs seems to be needed.

Reply:
When building during CI waiting ~20 minutes isn't the end of the world, IMO. I've added parameters to prune/limit the depth and nodes per graph. I think it's OK to leave this in the workflow, but could be convinced otherwise.

Reply:
Is this now done for every PR that gets submitted to stdlib? 

20 minutes is not the end of the world, but it significantly slows down the development.

Do you know why it takes so long?

Reply:
> Do you know why it takes so long?

I guess that the issue is the generation of the graphs for all generated (hundreds?) functions in the different submodules.

Reply:
It's fine for now, but if it already takes 20 minutes, I am a bit worried it will take hours once we have more functionality in stdlib.

Reply:
IMO, we shouldn't document specific interfaces at all, only generic ones (and the doc should list supported arguments and ranks). 

@zbeekman @cmacmackin does FORD have this capability?

Reply:
@certik @jvdp1 yes, in general the graphs and the search documentation take a while to generate. Installing the python LXML lib helps (in theory at least) with the search generation. Restricting graph depth and nodes might help with graph generation. Search generation can be easily and quickly disabled on the command line (I can open a PR) so that we don't build the search index unless it's a tagged release. Graph enable/disable has no global command line option yet, but a PR would be welcome. (Any help maintaining FORD would be welcome, AFAIK.)

@milancurcic You can [restrict the scope of what gets documented](https://github.com/fortran-lang/stdlib/pull/183/files#diff-6cd537095fc5651e430618f903fb71f9R13-R15) which can reduce the graph/search build time too. The only problem is that, as far as I know, if you do this FORD won't currently extract all of the specific interfaces for the generic. So we may have to document this manually with the FORD comment markup or open a PR to change how FORD behaves.

Should we try removing the private scope and see if that looks better? I can include that in a PR that turns of search index building unless it's being deployed.

Reply:
We can also limit MAXRANK to 3 as suggested by @jvdp1. Should speed things up, and folks can extrapolate to higher rank code.

Reply:
@zbeekman do you know how to restrict the rank? As it is now, all tests finish quickly except this documentation test that takes forever. Anything we can do to speed it up would help, at least in our initial stages. As we turn into production (we move functionality from experimental to main), we can revisit this.

Reply:
> @zbeekman do you know how to restrict the rank? As it is now, all tests finish quickly except this documentation test that takes forever. Anything we can do to speed it up would help, at least in our initial stages. As we turn into production (we move functionality from experimental to main), we can revisit this.

@certik Yes: [`MAXRANK=<n>` here](https://github.com/fortran-lang/stdlib/pull/183/files#diff-6cd537095fc5651e430618f903fb71f9R11) Do you want me to submit a PR restricting API documentation to public and protected entities only, reducing the MAXRANK and turning off search index and graph building unless documentation is going to be deployed? Also, what RANK should we restrict to? 2 or 3?

Reply:
If you could do that, that would be awesome. Can search index + graph building fail? If so, then we have to test it. Last thing we want is to fail at deploy.

Regarding rank, I don't know, try 3 and if it is fast enough, let's keep it.

Reply:
@certik 
> Can search index + graph building fail?

I've never seen search indexing fail and I don't think I've ever seen graph building fail, but this one I'm more nervous about.